cmd/teleproxy/teleproxy.go

```diff
@@ -185,98 +234,204 @@ func intercept(dnsIP string, fallbackIP string) (func(), error) {
 		} else {
 			fallbackIP = "8.8.8.8"
 		}
-		log.Printf("TPY: Automatically set -fallback=%v", dnsIP)
+		log.Printf("TPY: Automatically set -fallback=%v", fallbackIP)
 	}
 	if fallbackIP == dnsIP {
-		return nil, errors.New("if your fallbackIP and your dnsIP are the same, you will have a dns loop")
+		return errors.New("if your fallbackIP and your dnsIP are the same, you will have a dns loop")
 	}
 
 	iceptor := interceptor.NewInterceptor("teleproxy")
-
 	apis, err := api.NewAPIServer(iceptor)
 	if err != nil {
-		return nil, errors.Wrap(err, "API Server")
+		return errors.Wrap(err, "API Server")
 	}
 
-	srv := dns.Server{
-		Listeners: dnsListeners("1233"),
-		Fallback:  fallbackIP + ":53",
-		Resolve: func(domain string) string {
-			route := iceptor.Resolve(domain)
-			if route != nil {
-				return route.Ip
-			} else {
-				return ""
+	sup.Supervise(&supervisor.Worker{
+		Name:     "interceptor",
+		Requires: []string{}, // XXX: this will need to include the api server once it is changed to not bind early
+		Work: func(p *supervisor.Process) error {
+			iceptor.Start()
+			bootstrap := route.Table{Name: "bootstrap"}
+			bootstrap.Add(route.Route{
+				Ip:     dnsIP,
+				Target: "1233",
+				Proto:  "udp",
+			})
+			bootstrap.Add(route.Route{
+				Name:   "teleproxy",
+				Ip:     "127.254.254.254",
+				Target: apis.Port(),
+				Proto:  "tcp",
+			})
+
+			iceptor.Update(bootstrap)
+			p.Ready()
+			<-p.Shutdown()
+			iceptor.Stop()
+			return nil
+		},
+	})
+
+	sup.Supervise(&supervisor.Worker{
+		Name:     "api-server",
+		Requires: []string{},
+		Work: func(p *supervisor.Process) error {
+			apis.Start()
+			p.Ready()
+			<-p.Shutdown()
+			apis.Stop()
+			return nil
+		},
+	})
+
+	sup.Supervise(&supervisor.Worker{
+		Name:     "dns-server",
+		Requires: []string{},
+		Work: func(p *supervisor.Process) error {
+			srv := dns.Server{
+				Listeners: dnsListeners("1233"),
+				Fallback:  fallbackIP + ":53",
```

Comment:
What happens if DNS is not on `:53`?

Reply:
good question... my guess is stuff would break... ;-)
this has been hardcoded for a while though, so it doesn't seem to be that common in practice to use a non-standard port for dns... although who knows... I created another issue to track it